fix: truncate long system names

loujaybee opened the pull request:
Signed-off-by: Louis Bichard <contact@louisjohnbichard.co.uk>

## Update system diagram word wrapping

The current system diagram text element will exceed the parent element when the text is long. 

The change introduced truncates this text so that it doesn't exceed the parent. The fix isn't perfect, as it relies on a character count (which is not precise in width), however could be considered better than the current solution. 

The character count is set as `20`, but this is somewhat arbitrary. A more "comprehensive" fix could be to calculate text length, and break down the text in JS, but this might not be worth the complexity / added code, etc. 

Before:
![image](https://user-images.githubusercontent.com/5528307/121232901-f287e880-c889-11eb-9f7e-73349170d330.png)

After
![image](https://user-images.githubusercontent.com/5528307/121232865-e8fe8080-c889-11eb-80b1-53ebbce2a2f2.png)

Let me know if I've followed the contribution instructions correctly! A small fix right now, but hopefully more/bigger changes coming in future! 🚀

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

jhaals commented:
Thanks for the contribution @loujaybee 🙌 

Deep down inside I hoped to see`<marquee>` :trollface:

jhaals commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.tsx:
```diff
@@ -128,7 +128,9 @@ function RenderNode(props: DependencyGraphTypes.RenderNodeProps<any>) {
           style={{ fontWeight: 'bold' }}
         >
-          {props.node.name}
+          {props.node.name.length < 20
```
Probably nicest to declare this as `const name = ...length > 20: ? props.name` at the top with the other variables to make the render code a bit cleaner.

loujaybee replied:
Done! Hoisted it up, and pulled out the `20` int to a constant. Wasn't sure if you had any preferences on where the constants live, I chose to just keep it near to where it's used, but could be pulled up to the top of the module. 

```javascript
  const classes = useStyles();
  const catalogEntityRoute = useRouteRef(entityRouteRef);
  const kind = props.node.kind || 'Component';
  const ref = parseEntityRef(props.node.id);
  const MAX_NAME_LENGTH = 20;
  const truncatedNodeName = props.node.name.length < MAX_NAME_LENGTH ?
    props.node.name :
    `${props.node.name.slice(0, MAX_NAME_LENGTH)}...`
```

jhaals commented on .changeset/flat-chefs-push.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/plugin-catalog': patch
+---
+
+Truncate the entity name on the system diagram
```
Truncate long entity names on the system diagram

loujaybee replied:
Updated!

freben commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.tsx:
```diff
@@ -128,7 +128,9 @@ function RenderNode(props: DependencyGraphTypes.RenderNodeProps<any>) {
           alignmentBaseline="baseline"
           style={{ fontWeight: 'bold' }}
```
Have you tried if it works to use css ellipsis handling instead of manual truncation?

Something like

```ts
maxWidth: '200px',
textOverflow: 'ellipsis',
overflow: 'hidden',
whiteSpace: 'nowrap',
```

loujaybee replied:
I had a go with some CSS solutions, as it would be nicer, I agree. But I don't think this can apply since the `text` is part of the SVG, so I couldn't get it to apply wrapping CSS properties with an effect. 

Source: https://css-tricks.com/svg-properties-and-css/

loujaybee commented:
Amazing, thanks @jhaals / @freben for the feedback let me take a look and apply some updates!

loujaybee commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.test.tsx:
```diff
@@ -48,8 +48,8 @@ describe('<SystemDiagramCard />', () => {
       apiVersion: 'v1',
       kind: 'System',
       metadata: {
-        name: 'my-system2',
```
Removed these prefixes as a workaround, leaving these in would exceed the max character count, truncate and then the test would break. Updating these prefixes seemed like the least intrusive fix, rather than update assertions. I also considered having a default value within the function that applies the truncation, however it would need to be proxied through from the top level component and muddied some of the function signatures.

loujaybee commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.test.tsx:
```diff
@@ -127,7 +127,66 @@ describe('<SystemDiagramCard />', () => {
+
+    expect(getByText('namespace/alongdomai...')).toBeInTheDocument();
+    expect(getByText('namespace/alongsyste...')).toBeInTheDocument();
+    expect(getByText('namespace/alongentit...')).toBeInTheDocument();
```
Seems like a lot of boilerplate, but I guess it does have to mock the entities, etc.

loujaybee commented:
Thanks again @jhaals / @freben for the review, I've addressed the comments: added a unit test, extracted the truncation function code t to the top. Sadly I couldn't get it working with CSS, but I believe that's a SVG limitation. 

Do take a look and let me know if any further changes would be required! 

🚀🚀🚀

Rugvip reviewed (approved):
Looks good, thanks! 👍 

And thank you for adding tests, time to :shipit:

Rugvip commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.tsx:
```diff
@@ -73,10 +73,10 @@ function readableEntityName(
-        namespace?: string;
-        name: string;
-      },
+      kind: string;
+      namespace?: string;
+      name: string;
+    },
```
Some formatting trouble here

```suggestion
        kind: string;
        namespace?: string;
        name: string;
      },
```

loujaybee replied:
Thanks, updated !

Rugvip commented on plugins/catalog/src/components/SystemDiagramCard/SystemDiagramCard.tsx:
```diff
@@ -89,6 +89,10 @@ function RenderNode(props: DependencyGraphTypes.RenderNodeProps<any>) {
   const catalogEntityRoute = useRouteRef(entityRouteRef);
   const kind = props.node.kind || 'Component';
   const ref = parseEntityRef(props.node.id);
+  const MAX_NAME_LENGTH = 20;
+  const truncatedNodeName = props.node.name.length < MAX_NAME_LENGTH ?
+    props.node.name :
+    `${props.node.name.slice(0, MAX_NAME_LENGTH)}...`
```
```suggestion
  const MAX_NAME_LENGTH = 20;
  const truncatedNodeName =
    props.node.name.length < MAX_NAME_LENGTH
      ? props.node.name
      : `${props.node.name.slice(0, MAX_NAME_LENGTH)}...`;
```

loujaybee replied:
Thanks, updated!

Rugvip reviewed (commented):
Thanks!
